Secret generation improvements

jgustie opened the pull request:
Several improvements to secret generation highlighted by the current authorization mess. Again, the main goal is to unblock developers; however we found that the errors were not helpful (they still are pretty vague but we do not get anything better from the server right now).

jgustie committed:
Improve logic for default authorization client name

jgustie committed:
Do not include empty values in authorization secret

jgustie committed:
Allow configuration of explicit client ID/secret for secret generation

jgustie committed:
Change registration error message

It is more important to report what the server error was then the fact that the server sent the wrong type of content.

bradbeam commented on redskyctl/internal/commands/authorize_cluster/generator.go:
```diff
@@ -76,15 +76,31 @@ func NewGeneratorCommand(o *GeneratorOptions) *cobra.Command {
 	}
 
+	// Provide a more meaningful default client name if possible
+	if o.ClientName == "" {
```
is it worth moving this into `generate()` so we can `s/context.TODO()/cmd.Context()/`?

jgustie replied:
I put it here so we could include the answer as the default value for the CLI flag (i.e. it shows up in `--help`). The main thing attached to the `cmd.Context()` is the UA strings for HTTP requests which we don't need here. I used `TODO()` instead of `Background()` because the usage better fits the description.

That being said, when I moved it, it still uses `cfg.Kubectl` which is going to be problematic because we haven't hit `preRun` yet (nor will we for `--help`) so the `cfg` hasn't been loaded yet. I'll change it to exec `kubectl` directly in this case.

bradbeam commented on internal/oauth2/registration/registration.go:
```diff
@@ -175,10 +175,9 @@ func doRegistrationRoundTrip(ctx context.Context, req *http.Request, src oauth2.
-			}
+		if t, _, err := mime.ParseMediaType(r.Header.Get("Content-Type")); err != nil || t != "application/json" {
+			return nil, fmt.Errorf("registration: unexpected server error (%v - %s)", code, http.StatusText(code))
+		} else if err := json.Unmarshal(body, responseError); err != nil {
```
this seems like it could be broken up to help readability -- 

```
if t, _, err := mime.ParseMediaType(r.Header.Get("Content-Type")); err != nil || t != "application/json" {
 			return nil, fmt.Errorf("registration: unexpected server error (%v - %s)", code, http.StatusText(code))
}

if err := json.Unmarshal(body, responseError); err != nil {
...
```

jgustie committed:
Fix secret generation default client name

The configuration has not been loaded yet when client name needs to be initialized (and in the case of --help it never will be).

Directly exec'ing the command removes the dependency on loading the configuration.

Also, added a KUBECONFIG when generating documentation to prevent the local environment from leaking into the published documentation.

jgustie committed:
Code review
